File: Java/160.Intersection_of_Two_Linked_Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
import java.util.HashSet;

public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        Set<ListNode> a_nodes = new HashSet<>();
        while (headA != null) {
            a_nodes.add(headA);
            headA = headA.next;
        }
        while (headB != null) {
            if (a_nodes.contains(headB)) {
                return headB;
            }
            headB = headB.next;
        }
        return null;
    }
}

File: Python3/160.Intersection_of_Two_Linked_Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        a_nodes = set();
        while headA:
            a_nodes.add(headA)
            headA = headA.next
        
        while headB:
            if headB in a_nodes:
                return headB
            headB = headB.next
        

